MergerService/src/run.cs

```diff
@@ -51,153 +59,252 @@ private List<IData> BuildDataList(Source[] paths, int batchSize)
-                }
-                catch (Exception e)
+                foreach (var item in taskTypes)
                 {
-                    this._logger.LogError($"Error in MergerService run - get task: {e.Message}");
+                    string jobType = item.Key;
+                    string taskType = item.Value;
+
+                    try
+                    {
+                        task = taskUtils.GetTask(jobType, taskType);
+                    }
+                    catch (Exception e)
+                    {
+                        this._logger.LogError($"Error in MergerService start - get task: {e.Message}");
+                    }
+
+                    // Guard clause in case there are no batches or sources
+                    if (task == null)
+                    {
+                        continue;
+                    }
+
+                    RunTask(task, taskUtils);
+                    activatedAny = true;
+
+                    try
+                    {
+                        taskUtils.UpdateCompletion(task.JobId, task.Id);
+                        taskUtils.NotifyOnCompletion(task.JobId, task.Id);
+                        this._logger.LogInformation($"Completed task: jobId: {task.JobId}, taskId: {task.Id}");
+                    }
+                    catch (Exception e)
+                    {
+                        this._logger.LogError($"Error in MergerService start - update task: {e.Message}");
+                    }
                 }
 
-                // Guard clause in case there are no batches or sources
-                if (task == null || task.Batches == null || task.Sources == null)
+                // Sleep only if there was no task to run for any type
+                if (!activatedAny)
                 {
-                    // Sleep for 0.5 sec so we won't send too many requests if there are no tasks available
-                    // TODO: read from configuration
-                    Thread.Sleep(1000);
-                    continue;
+                    // Sleep for 1 sec so we won't send too many requests if there are no tasks available
+                    Thread.Sleep(pollingTime);
                 }
+            }
+        }
+
+        private void RunTask(MergeTask task, ITaskUtils taskUtils)
+        {
+            // Guard clause in case there are no batches or sources
+            if (task.Parameters is null || task.Parameters.Batches is null || task.Parameters.Sources is null)
+            {
+                this._logger.LogWarning($"Task parameters are invalid. Task: {task.ToString()}");
+                taskUtils.UpdateReject(task.JobId, task.Id, task.Attempts, "Task parameters are invalid", false);
+                return;
+            }
 
-                // Log the task
-                this._logger.LogInformation($"starting task: taskId job: jobId");//TODO: add job and task ids
+            bool validate = this._configurationManager.GetConfiguration<bool>("GENERAL", "validate");
```

Comment:
why get in each task and no once on start or ctor?

Reply:
Moved to constructor.